src/mpid/ch4/shm/ipc/gpu/gpu_post.c

```diff
@@ -398,13 +417,21 @@ int MPIDI_GPU_fill_ipc_handle(MPIDI_IPCI_ipc_attr_t * ipc_attr,
         mpi_errno = ipc_track_cache_search(track_tree, pbase, len, &handle, &found);
         MPIR_ERR_CHECK(mpi_errno);
 
-        if (found) {
+        if (found && MPL_gpu_ipc_handle_is_valid(&handle, pbase)) {
             handle_status = MPIDI_GPU_IPC_HANDLE_VALID;
             goto fn_done;
+        } else if (found) {
+            /* remove and destroy invalid handle */
+            mpi_errno = ipc_track_cache_remove(pbase, len, local_dev_id);
+            MPIR_ERR_CHECK(mpi_errno);
+
+            mpl_err = MPL_gpu_ipc_handle_destroy(pbase, &ipc_attr->u.gpu.gpu_attr);
+            MPIR_ERR_CHKANDJUMP(mpl_err != MPL_SUCCESS, mpi_errno, MPI_ERR_OTHER,
+                                "**gpu_ipc_handle_destroy");
```

Comment:
How about restructure the if-chain as -
```
if (found) {
    if (MPL_gpu_ipc_handle_is_valid(...)){
        ...
        goto fn_done;
    } else {
        /* remove and destroy */
        ...
    }
}
```
?

I find a non-terminating `else if` less easy to parse.
```

Reply:
will do

Reply:
done